src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -158,6 +222,140 @@ async def generate_lockfile(
+        PathGlobs(
+            [req.lockfile_dest],
+            glob_match_error_behavior=GlobMatchErrorBehavior.ignore,
+        ),
+    )
+    if maybe_lockfile.serialized_bytes_length == 0:
+        raise ValueError(
+            f"Empty or non-existent lockfile at {req.lockfile_dest}. Unable to update in place"
+        )
+
+    original_loaded_lockfile = await Get(
+        LoadedLockfile,
+        LoadedLockfileRequest(
+            Lockfile(
+                url=req.lockfile_dest,
+                url_description_of_origin=f"existing lockfile from {req.resolve_name}",
+                resolve_name=req.resolve_name,
+            )
+        ),
+    )
+
+    if not (
+        original_loaded_lockfile.metadata
+        and isinstance(original_loaded_lockfile.metadata, (PythonLockfileMetadataV3))
+    ):
+        raise ValueError(
+            f"Pants metadata for lockfile at {req.lockfile_dest} is in old <V3 format. Unable to update in place."
+        )
+    elif set(resolve_config.no_binary) != original_loaded_lockfile.metadata.no_binary:
+        raise ValueError(
+            f"Resolve Config no_binary {set(resolve_config.no_binary)} does not match {original_loaded_lockfile.metadata.no_binary} in last lockfile, can not update in place"
+        )
+    elif set(resolve_config.only_binary) != original_loaded_lockfile.metadata.only_binary:
+        raise ValueError(
+            f"Resolve Config only_binary {set(resolve_config.only_binary)} does not match {original_loaded_lockfile.metadata.only_binary} in last lockfile, can not update in place"
+        )
+
+    # Make sure to keep Pant's record of the specifier as the source of truth
+    # and not introduce inconsistencies with CLI arguments
+    requirements_projects_to_update: List[str] = []
+    requirements_projects = {
+        PipRequirement.parse(requirement).project_name: requirement
+        for requirement in req.requirements
+    }
+    for project in pex_lock_subsystem.project:
+        if project not in requirements_projects:
+            raise ValueError(
+                f"project {project} not found among known requirements. Projects msut be listed in known Pants targets"
+            )
+        requirements_projects_to_update.append(requirements_projects[project])
+    if len(requirements_projects_to_update):
+        logger.info(f"Updating requirements for lockfile: {requirements_projects_to_update}")
+
+    inferred_new_projects: List[str] = []
+    for maybe_new_requirement in req.requirements:
+        maybe_new_pip_req = PipRequirement.parse(maybe_new_requirement)
+        if (
+            maybe_new_pip_req not in original_loaded_lockfile.metadata.requirements
+            and str(maybe_new_requirement) not in requirements_projects_to_update
+        ):
+            inferred_new_projects.append(maybe_new_requirement)
+    if len(inferred_new_projects) > 0:
+        logger.info(f"Inferred new requirements for lockfile update: {inferred_new_projects}")
```

Comment:
Reading this makes me wonder what happens if a requirement is removed? Does PEX support doing an update that removes requirements from the lockfile? Potentially we should identify them, and either pass them to pex (if supported) or emit an error/warning?

Reply:
> Does PEX support doing an update that removes requirements from the lockfile?

It does not today. That seems a reasonable thing to support though.

Reply:
This is now available as `-d <project name>` / `--delete-project <project name>` (#2335) released in [Pex 2.1.160](https://github.com/pantsbuild/pex/releases/tag/v2.1.160).